### PR TITLE
Content export

### PR DESCRIPTION
**Pull request by jefferya**
First version of the content export for transition to Aviary (creates base to work from) - addresses #753 

Exports:
* collection with metadata
* media object with
  *  metadata
  * MODS
  * structural metadata
  * media file
  * derivatives
  * captions
  * thumbnail
  * poster
  
See #753 for details

### CODE REVIEW

File: app/controllers/media_objects_controller.rb [hunk 1]
```diff
@@ -336,7 +336,7 @@ def show
       format.json do
         response_json = @media_object.as_json
         response_json.except!(:files, :visibility, :read_groups) unless current_ability.is_administrator?
-        render json: response_json.to_json
+        render json: response_json.to_json(:except => :captions)
```
Comment (jefferya) [inline app/controllers/media_objects_controller.rb:339]: Some caption files not UTF-8 therefore causes failures only in the JSON output format (i.e., works with media player). This API is not used. Content clean-up will happen after export. Captions can be exported in a different manner.

File: app/models/concerns/derivative_intercom.rb [hunk 2]
```diff
@@ -16,7 +16,7 @@ module DerivativeIntercom
   def to_ingest_api_hash
     {
       label: "quality-#{quality}", # quality-low, quality-medium, quality-high
-      id: track_id,
+      id: id,
```
Comment (jefferya) [inline app/models/concerns/derivative_intercom.rb:19]: Bugfix from upstream (id wrong value --upstream avalonmediasystem@f6ec5bd)

File: app/models/concerns/master_file_intercom.rb [hunk 3]
```diff
@@ -15,6 +15,7 @@
 module MasterFileIntercom
   def to_ingest_api_hash(include_structure = true)
     {
+      id: id,
```
Comment (jefferya) [inline app/models/concerns/master_file_intercom.rb:18]: Adds missing `id` field

File: lib/tasks/avalon_ualbertalib.rake [hunk 4]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+        obj = ActiveFedora::Base.find(options[:id])
+        if (obj.is_a?(Admin::Collection))
+          obj.media_objects.to_a.collect.each do |m| 
+            puts "%s" % [m.id] 
+          end;
+        else
+          puts "ERROR: not a Collection obj: [%s]" % [options[:id]] 
+        end
+
+      end
```
Comment (jefferya) [inline lib/tasks/avalon_ualbertalib.rake:109]: Returns a list of ids representing all members of a specified collection
`bundle exec rake avalon:ualbertalib:list_collection_members -- --collection_id=mc87pq24j`

File: lib/tasks/avalon_ualbertalib.rake [hunk 5]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+          File.open("#{poster_base_dir}/#{mf.poster.original_name}", 'w:BINARY') { |file| file.write(mf.poster.content) }
+        end
+        if mf.thumbnail.has_content?
+          thumbnail_base_dir = File.join(mf_base_dir,"thumbnail")
+          FileUtils.mkdir_p(thumbnail_base_dir)
+          File.open("#{thumbnail_base_dir}/#{mf.thumbnail.original_name}", 'w:BINARY') { |file| file.write(mf.thumbnail.content) }
+        end
+      end
+
     end
```
Comment (jefferya) [inline lib/tasks/avalon_ualbertalib.rake:243]: Creates version 1 of the export package:

`bundle exec rake avalon:ualbertalib:export_contents -- --objects="/tmp/ids.txt" --dest_base="/tmp/delete_dest/" -derivatives="masterfiles/delete/`

File: lib/tasks/avalon_ualbertalib.rake [hunk 6]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+          puts "You must specify an set of MediaObject IDs."
+          puts option_parser.help
+          exit 1
+        end
+      end
+
+      def ualberta_export_collectionObject( obj, options )
+        base_dir = File.join(options[:dest_base_dir], obj.id)
+        FileUtils.mkdir_p(base_dir)
+        File.open("#{base_dir}/collection_object.json", 'w') { |file| file.write(obj.as_json.to_json(:except => :captions)) }
```
Comment (pgwillia) [inline lib/tasks/avalon_ualbertalib.rake:177]: The hash rocket (`key => value`) syntax was common before Ruby 1.9.  I think the preferred form is now `key: value`
```suggestion
        File.open("#{base_dir}/collection_object.json", 'w') { |file| file.write(obj.as_json.to_json(except: :captions)) }
```

File: lib/tasks/avalon_ualbertalib.rake [hunk 7]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+      end
+
+      def ualberta_export_mediaObject( obj, options )
+        base_dir = File.join(options[:dest_base_dir], obj.collection_id, "MediaObjects", obj.id)
+        FileUtils.mkdir_p(base_dir)
+        File.open("#{base_dir}/media_object.json", 'w') { |file| file.write(obj.as_json.to_json(:except => :captions)) }
```
Comment (pgwillia) [inline lib/tasks/avalon_ualbertalib.rake:183]: I'm wondering if `as_json.to_json` is redundant?  Could it just be `to_json`?
```suggestion
        File.open("#{base_dir}/media_object.json", 'w') { |file| file.write(obj.to_json(:except => :captions)) }
```

File: lib/tasks/avalon_ualbertalib.rake [hunk 8]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+        File.open("#{base_dir}/mods.xml", 'w:BINARY') { |file| file.write(obj.descMetadata.content) }
+        obj.ordered_master_files.to_a.collect.each do |mf|
```
Comment (pgwillia) [inline lib/tasks/avalon_ualbertalib.rake:185]: Another instance of the `to_a.collect.each` pattern.

File: lib/tasks/avalon_ualbertalib.rake [hunk 9]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+          # log: if media doesn't exist or has been moved
+          glob = Dir.glob("#{ENV['SETTINGS__DROPBOX__PATH']}/**/#{File.basename(mf.file_location)}")
+          puts "ERROR: [%s] media file not found: [%s] for Object [%s]" % [mf.id, mf.file_location, mf.media_object_id] if glob.empty?
+          # assumption: every media file should have a unique name otherwise this fails and fails by reporting 2+ items
+          # log: if multiple files of the same name found but don't change execution path as there could be a legitemate reason
+          puts "ERROR: [%s] stored media file path outdated: multiple media files found with name [%s]" % [mf.id, File.basename(mf.file_location) ] if glob.count > 1
+          glob.each do |file|
+            FileUtils.mkdir_p(original_base_dir)
+            FileUtils.cp(file, "#{original_base_dir}/")
+            puts "INFO: [%s] stored media file path outdated [%s] -- using alternaive [%s]" % [mf.id, mf.file_location, file]
```
Comment (pgwillia) [inline lib/tasks/avalon_ualbertalib.rake:211]:
```suggestion
            puts "INFO: [%s] stored media file path outdated [%s] -- using alternative [%s]" % [mf.id, mf.file_location, file]
```

File: lib/tasks/avalon_ualbertalib.rake [hunk 10]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
...
+          end
+        end
+        mf.derivatives.to_a.collect.each do |derivative|
```
Comment (pgwillia) [inline lib/tasks/avalon_ualbertalib.rake:214]: More `to_a.collect.each` pattern.

**Review by pgwillia — DISMISSED** [review]
I think with this we want to move with speed so I'm good with this as long as it's working.

I do have some comments about how to simplify some parts.  I hope that these kinds of comments are helpful to you.

File: app/controllers/media_objects_controller.rb (hunk 1 above)
Comment (jefferya) [reply app/controllers/media_objects_controller.rb:339]: The `unless` slightly changes the scope, would prefer have the two exceptions sets separate with the `captions` covering the entire scope. Feels wrong to add tech debt but the real solution is to fix the need behind including captions here (white screen of death when caption files include non utf-8 characters).
```
#line 338
response_json.except!(:files, :visibility, :read_groups) unless current_ability.is_administrator?
```

File: lib/tasks/avalon_ualbertalib.rake [hunk 11]
```diff
@@ -66,6 +68,177 @@ namespace :avalon do
           exit 1
         end
       end
-      
+
+      desc "Return a list of collections: id's and labels"
+      task :list_collections => :environment do
+
+        Admin::Collection.all.to_a.collect.each do |collection|
```
[the remark this answers is not in the document]
Comment (jefferya) [reply lib/tasks/avalon_ualbertalib.rake]: Thanks! I didn't know about find_each.

**Review by pgwillia — APPROVED** [review]
👍 Looks good!

**Pull request merged**